### PR TITLE
[vm] Implement push, pop, borrow for native vector.

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Implement push_back, pop, and borrow for native vectors. This will give us the MVP for collection functionality needed by the validator set contracts. I added one error code to represent errors in the Native Functions, which was treated as abort previously, which thus changed the tests for the signature checking primitive tests.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added tests for all runtime errors for vector.

## Related PRs

#869

### CODE REVIEW

File: language/stdlib/modules/vector.mvir [hunk 1]
```diff
@@ -16,6 +16,12 @@ module Vector {
   // Add an element to the end of the vector
   native public push_back<Element>(v: &mut Self.T<Element>, e: Element);
 
+  // Get the mutable reference to the nth element in the vector, abort if out of bound.
```
Comment (sblackshear) [inline language/stdlib/modules/vector.mvir]:
```suggestion
  // Get a mutable reference to the ith element in the vector, abort if out of bounds.
```

File: language/stdlib/modules/vector.mvir [hunk 2]
```diff
@@ -16,6 +16,12 @@ module Vector {
...
+  native public borrow_mut<Element>(v: &mut Self.T<Element>, idx: u64): &mut Element;
+
+  // Pop an element from the vector. abort if the vector is empty.
```
Comment (sblackshear) [inline language/stdlib/modules/vector.mvir]:
```suggestion
  // Pop an element from the end of vector, abort if the vector is empty.
```

File: language/stdlib/modules/vector.mvir [hunk 3]
```diff
@@ -16,6 +16,12 @@ module Vector {
...
+  native public pop<Element>(v: &mut Self.T<Element>): Element;
```
Comment (sblackshear) [inline language/stdlib/modules/vector.mvir]: Nit: name this `pop_back` so it's clear which end we pop from?

File: language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs [hunk 4]
```diff
@@ -46,46 +54,87 @@ impl NativeVector {
...
+                error_code: POP_EMPTY_VEC,
+            },
+            Some(Ok(v)) => {
+                println!("Popped {:?}", v);
+                NativeReturnStatus::Success {
+                    cost: POP_COST,
+                    return_values: vec![v],
+                }
+            }
+            // The poped element has dangling references.
```
Comment (sblackshear) [inline language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]:
```suggestion
            // The popped element has dangling references.
```

Comment (sblackshear) [reply language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]: When can this happen? Should it be prevented by the bytecode verifier?

File: language/functional_tests/tests/testsuite/natives/vector/vector_module.mvir [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+//! new-transaction
+import {{default}}.M;
+import 0x0.LibraAccount;
+import 0x0.LibraCoin;
+
+main() {
+  let coin: LibraCoin.T;
```
Comment (vgao1996) [inline language/functional_tests/tests/testsuite/natives/vector/vector_module.mvir]: Indents in tests are inconsistent. Could you make all of them either 2 or 4 spaces?

File: language/functional_tests/tests/testsuite/natives/vector/vector_pop_out_of_range.mvir [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    v = Vector.empty<u64>();
+    value = Vector.pop<u64>(&mut v);
+    return;
+  }
+}
+
+//! new-transaction
+import {{default}}.M;
+main() {
+    M.pop_out_of_range();
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/natives/vector/vector_pop_out_of_range.mvir]: Why not get rid of the module and inline the code from `pop_out_of_range` here?

File: language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs [hunk 7]
```diff
@@ -46,46 +54,87 @@ impl NativeVector {
...
+        match reference.mutate_native_struct(|native_val| {
+            get_mut_vector(native_val)?.0.pop().map(MutVal::into_value)
+        }) {
+            // Vector is already empty.
+            None => NativeReturnStatus::Aborted {
+                cost: POP_COST,
+                error_code: POP_EMPTY_VEC,
+            },
+            Some(Ok(v)) => {
+                println!("Popped {:?}", v);
```
Comment (sblackshear) [inline language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]: Delete?

File: types/src/vm_error.rs [hunk 8]
```diff
@@ -469,6 +469,7 @@ pub enum StatusCode {
     CODE_DESERIALIZATION_ERROR = 4019,
     EXECUTION_STACK_OVERFLOW = 4020,
     CALL_STACK_OVERFLOW = 4021,
+    NATIVE_FUNCTION_ERROR = 4022,
```
Comment (sblackshear) [inline types/src/vm_error.rs]: What are we currently doing for errors in (e.g.) native crypto functions?

Comment (runtian-zhou) [reply types/src/vm_error.rs]: That was exactly what I did. Previously the native crypto functions are treated as assertion failures. The returned status code will be `ABORTED`. It seems to me that it is not ideal as we will conflate the errors code namespace between assertions failures and native function failures cc @tzakian

File: language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs (hunk 4 above)
Comment (runtian-zhou) [reply language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]: Exactly. This should be prevented by the verifier.

Comment (sblackshear) [reply language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]: Got it, thanks! Should we add a special `InternalError` error code here like we do for the other errors that we expect to be impossible due to the verifier?

Comment (sblackshear) [reply language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs]: Discussed offline; `InvalidArguments` seems to be a catchall type for internal errors that should be prevented by the bytecode verifier.

**Pull request closed without merging**